packages/lib/tag/auth.ts

```diff
@@ -0,0 +1,26 @@
+import { hasUserEnvironmentAccessCached } from "../environment/auth";
+import { getTag } from "./service";
+import { unstable_cache } from "next/cache";
+
+export const canUserAccessTag = async (userId: string, tagId: string): Promise<boolean> => {
+  if (!userId) return false;
+
+  const tag = await getTag(tagId);
+  if (!tag) return false;
+
+  const hasAccessToEnvironment = await hasUserEnvironmentAccessCached(userId, tag.environmentId);
+  if (!hasAccessToEnvironment) return false;
+
+  return true;
+};
+
+export const canUserAccessTagCached = async (userId: string, tagId: string) =>
```

Comment:
Please build the cache directly into canUserAccessTag. Choose `users-${userId}-tags-${tagId}` as the key and the tag as a tag ;-) `tags-${tagId}`